homeassistant/components/mqtt/__init__.py

```diff
@@ -130,25 +130,45 @@
     CONF_WILL_MESSAGE,
 ]
 
+_T = TypeVar("_T")
+
+REMOVED_OPTIONS = vol.All(
+    cv.removed(CONF_BIRTH_MESSAGE),  # Removed in HA Core 2023.4
+    cv.removed(CONF_BROKER),  # Removed in HA Core 2023.4
+    cv.removed(CONF_CERTIFICATE),  # Removed in HA Core 2023.4
+    cv.removed(CONF_CLIENT_ID),  # Removed in HA Core 2023.4
+    cv.removed(CONF_CLIENT_CERT),  # Removed in HA Core 2023.4
+    cv.removed(CONF_CLIENT_KEY),  # Removed in HA Core 2023.4
+    cv.removed(CONF_DISCOVERY),  # Removed in HA Core 2022.3
+    cv.removed(CONF_DISCOVERY_PREFIX),  # Removed in HA Core 2023.4
+    cv.removed(CONF_KEEPALIVE),  # Removed in HA Core 2023.4
+    cv.removed(CONF_PASSWORD),  # Removed in HA Core 2023.4
+    cv.removed(CONF_PORT),  # Removed in HA Core 2023.4
+    cv.removed(CONF_PROTOCOL),  # Removed in HA Core 2023.4
+    cv.removed(CONF_TLS_INSECURE),  # Removed in HA Core 2023.4
+    cv.removed(CONF_USERNAME),  # Removed in HA Core 2023.4
+    cv.removed(CONF_WILL_MESSAGE),  # Removed in HA Core 2023.4
+)
+
+
+def _warn_config_style(config: list[Any] | _T) -> list[Any] | _T:
+    """Detect not conformant yaml configuration."""
+    if not isinstance(config, list):
+        _LOGGER.debug(
+            "Your MQTT configuration format does not comply ADR 0007. Expected a list, got %s. See also https://github.com/home-assistant/architecture/discussions/906",
+            type(config),
+        )
+    return config
+
+
```

Comment:
I'm not quite sure we need this check, shouldn't we just add it if we decide to deprecate the non-list format some time in the future?